core/src/main/java/com/linecorp/armeria/server/websocket/WebSocketService.java

```diff
@@ -0,0 +1,291 @@
+        final WebSocket outboundFrames = handler.handle(ctx, new WebSocketWrapper(inboundFrames));
+        decoder.setOutboundWebSocket(outboundFrames);
+        return HttpResponse.of(
+                responseHeaders, outboundFrames.recoverAndResume(cause -> {
+                                                   if (cause instanceof ClosedStreamException) {
+                                                       return StreamMessage.aborted(cause);
+                                                   }
+                                                   ctx.logBuilder().responseCause(cause);
+                                                   return StreamMessage.of(newCloseWebSocketFrame(cause));
+                                               })
+                                               .map(frame -> HttpData.wrap(encoder.encode(ctx, frame))));
+    }
+
+    /**
+     * Handles the HTTP/2 web socket handshake described in
+     * <a href="https://datatracker.ietf.org/doc/html/rfc8441">Bootstrapping WebSockets with HTTP/2</a>.
+     * These are examples of a request and the corresponding response:
+     *
+     * <p>Request:
+     * <pre>
+     * HEADERS + END_HEADERS
+     * :method = CONNECT
+     * :protocol = websocket
+     * :scheme = https
+     * :path = /chat
+     * :authority = server.example.com
+     * sec-websocket-protocol = chat, superchat
+     * sec-websocket-extensions = permessage-deflate
+     * sec-websocket-version = 13
+     * origin = http://www.example.com
+     * </pre>
+     *
+     * <p>Response:
+     * <pre>
+     * HEADERS + END_HEADERS
+     * :status = 200
+     * sec-websocket-protocol = chat
+     * </pre>
+     */
+    @Override
+    protected HttpResponse doConnect(ServiceRequestContext ctx, HttpRequest req) throws Exception {
+        if (!ctx.sessionProtocol().isExplicitHttp2()) {
+            return HttpResponse.of(HttpStatus.METHOD_NOT_ALLOWED);
+        }
+        final RequestHeaders headers = req.headers();
+        if (!isHttp2WebSocketUpgradeRequest(headers)) {
+            logger.trace("RequestHeaders does not contain headers for WebSocket upgrade. headers: {}", headers);
+            return missingHttp2WebSocketUpgradeHeader.toHttpResponse();
+        }
+
+        final HttpResponse invalidResponse = checkOriginAndVersion(headers);
+        if (invalidResponse != null) {
+            return invalidResponse;
+        }
+
+        // As described in https://datatracker.ietf.org/doc/html/rfc8441#section-5,
+        // HTTP/2 does not use Sec-WebSocket-Key and Sec-WebSocket-Accept headers.
+
+        final ResponseHeadersBuilder responseHeadersBuilder = ResponseHeaders.builder(HttpStatus.OK);
+        maybeAddSubprotocol(headers, responseHeadersBuilder);
+        return handleUpgradeRequest(ctx, req, responseHeadersBuilder.build());
+    }
+
+    @Nullable
+    private HttpResponse checkOriginAndVersion(RequestHeaders headers) {
+        if (!allowedOrigins.isEmpty()) {
```

Comment:
The default value of `allowedOrigins` is empty and browsers do not enforce the [Same Origin Policy](https://en.wikipedia.org/wiki/Same-origin_policy) for WebSocket connections. That means users need to explicitly protect their services without sensible default protection.

If `allowedOrigins` are not set, how about restricting the origin to the current authority (domain)? We may compare `Origin` header with `:authority` header after removing the scheme part (`<scheme>://`) from the origin header.

Reference: 
- https://docs.spring.io/spring-security/reference/servlet/integrations/websocket.html#websocket-sameorigin-spring
- https://en.wikipedia.org/wiki/Same-origin_policy
- https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Origin

Reply:
That's a good point. Let me use the way you suggested. 
I also found a relevant article: 
https://christian-schneider.net/CrossSiteWebSocketHijacking.html